runtime/src/accounts.rs

```diff
@@ -1080,102 +1055,99 @@ impl Accounts {
-    maybe_nonce: Option<(
-        &Pubkey,
-        &AccountSharedData,
-        Option<&AccountSharedData>,
-        bool,
-    )>,
+    is_fee_payer: bool,
+    maybe_nonce: Option<&'a NonceFull>,
     blockhash: &Hash,
     lamports_per_signature: u64,
 ) -> bool {
-    if let Some((nonce_key, nonce_account, _, advance_blockhash)) = maybe_nonce {
-        if address == nonce_key {
-            if execution_result.is_err() {
-                // The transaction failed which would normally drop the account
-                // processing changes, since this account is now being included in
-                // the accounts written back to the db, roll it back to
-                // pre-processing state.
-                *account = nonce_account.clone();
+    if let Some(nonce) = maybe_nonce {
+        if address == nonce.address() {
+            // The transaction failed which would normally drop the account
+            // processing changes, since this account is now being included
+            // in the accounts written back to the db, roll it back to
+            // pre-processing state.
+            *account = nonce.account().clone();
+
+            // Advance the stored blockhash to prevent fee theft by someone
+            // replaying nonce transactions that have failed with an
+            // `InstructionError`.
+            //
+            // Since we know we are dealing with a valid nonce account,
+            // unwrap is safe here
+            let state = StateMut::<NonceVersions>::state(nonce.account())
+                .unwrap()
+                .convert_to_current();
+            if let NonceState::Initialized(ref data) = state {
+                account
+                    .set_state(&NonceVersions::new_current(NonceState::new_initialized(
+                        &data.authority,
+                        blockhash,
+                        lamports_per_signature,
+                    )))
+                    .unwrap();
             }
-
-            if advance_blockhash {
-                // Advance the stored blockhash to prevent fee theft by someone
-                // replaying nonce transactions that have failed with an
-                // `InstructionError`.
-                //
-                // Since we know we are dealing with a valid nonce account, unwrap is safe here
-                let state = StateMut::<NonceVersions>::state(nonce_account)
-                    .unwrap()
-                    .convert_to_current();
-                if let NonceState::Initialized(ref data) = state {
-                    account
-                        .set_state(&NonceVersions::new_current(NonceState::new_initialized(
-                            &data.authority,
-                            blockhash,
-                            lamports_per_signature,
-                        )))
-                        .unwrap();
+            true
+        } else {
+            if
+            /*TODO remove this first part? */
```

Comment:
As in, do this regardless of the result?

Reply:
Nah, I was playing around with the idea but forget to remove the todo, removed